Add attenuated MIP (attenuated_mip) rendering mode to `VolumeVisual`

alisterburt opened the pull request:
This PR upstreams the attenuated maximum intensity projection rendering mode from napari/napari#846 (by @sofroniewn) into vispy

djhoese commented on vispy/visuals/volume.py:
```diff
@@ -749,6 +779,17 @@ def threshold(self, value):
+    @attenuation.setter
+    def attenuation(self, value):
+        self._attenuation = float(value)
+        self.shared_program['u_attenuation'] = self._attenuation
```
This should probably have the `if 'u_attenuation' ...` like threshold's setter has.

djhoese commented on vispy/visuals/volume.py:
```diff
@@ -728,6 +756,8 @@ def method(self, method):
         # Get rid of specific variables - they may become invalid
         if 'u_threshold' in self.shared_program:
             self.shared_program['u_threshold'] = None
+        elif 'u_attenuation' in self.shared_program:
```
Can this maybe be made an `if` instead of `elif` for future flexibility? Otherwise this assumes that two shaders will never use both parameters. I know, nit-pick, but still.

djhoese reviewed (changes requested):
Just some small changes, but otherwise looks pretty good.

sofroniewn commented:
This is looking great @alisterburt! It will let us remove our vendored copy of the volume visual from napari

almarklein commented on vispy/visuals/volume.py:
```diff
@@ -707,8 +737,6 @@ def method(self):
               the result is opaque.
             * mip: maxiumum intensity projection. Cast a ray and display the
               maximum value that was encountered.
-            * minip: minimum intensity projection. Cast a ray and display the
-              minimum value that was encountered.
```
Why are these lines removed?

alisterburt replied:
whoops - merge shenanigans, will fix!

almarklein reviewed (commented):
Nice!

alisterburt commented:
thanks for the reviews @djhoese, @almarklein and @sofroniewn - much appreciated

I've updated to incorporate suggested changes, rebased onto master and fixed docstrings, assuming all tests pass I think we're good to go!

djhoese commented on vispy/visuals/volume.py:
```diff
@@ -731,6 +762,11 @@ def method(self):
               maximum value that was encountered.
             * minip: minimum intensity projection. Cast a ray and display the
               minimum value that was encountered.
+            * attenuated_mip: attenuated maximum intensity projection. Cast a
+              ray and display the maximum value encountered. Values are
+              attenuated as the ray moves deeper into the volume.
+            * minip: minimum intensity projection. Cast a ray and display the
+              minimum value that was encountered.
             * additive: voxel colors are added along the view ray until
```
```suggestion
            * attenuated_mip: attenuated maximum intensity projection. Cast a
              ray and display the maximum value encountered. Values are
              attenuated as the ray moves deeper into the volume.
            * additive: voxel colors are added along the view ray until
```

djhoese reviewed (approved):
I fixed one small thing (minip was listed twice in the docstring). Looks good now. It bothers me a little that `_attenuation` isn't directly declared in the `__init__` but threshold works the same way so :shrug: we'll ignore it for now. Thanks.

alisterburt commented:
thanks @djhoese for handling this so smoothly!
